core/server/master/src/main/java/alluxio/master/file/meta/InvalidationSyncCache.java

```diff
@@ -0,0 +1,413 @@
+ * - isFile: when checking if a file needs to be synced, any descendant type from the most
+ * recent sync will be valid
+ *
+ * Whenever an invalidation is received the path, and its parents up to the root have their
+ * appropriate invalidation times updated. Validation times are updated on the path
+ * when {@link #notifySyncedPath} is called on the root sync path after a successful sync.
+ *
+ * Checking if a path needs to be synchronized involves checking the appropriate validation
+ * and invalidation times at each path component up to the root.
+ */
+@ThreadSafe
+public class InvalidationSyncCache {
+  private static final Logger LOG = LoggerFactory.getLogger(InvalidationSyncCache.class);
+
+  public final Cache<String, SyncState> mItems;
+  private final Clock mClock;
+  // we keep the root as a separate value as we never want to evict it,
+  // and instead we just update it under a lock
+  private final SyncState mRoot = new SyncState(false);
+  private final Lock mRootLock = new ReentrantLock();
+
+  /**
+   * Creates a new instance of {@link InvalidationSyncCache}.
+   * @param clock the clock to use to compute sync times
+   */
+  public InvalidationSyncCache(Clock clock) {
+    this(clock, null);
+  }
+
+  @VisibleForTesting
+  InvalidationSyncCache(Clock clock,
+      @Nullable BiConsumer<String, SyncState> onRemoval) {
+    mClock = clock;
+    mItems = CacheBuilder.newBuilder()
+        .removalListener(
+            (removal) -> {
+              if (removal.wasEvicted() && removal.getKey() != null && removal.getValue() != null) {
+                if (onRemoval != null) {
+                  onRemoval.accept((String) removal.getKey(), (SyncState) removal.getValue());
+                }
+                onCacheEviction((String) removal.getKey(), (SyncState) removal.getValue());
+              }
+            })
+        .maximumSize(Configuration.getInt(
+            PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY))
+        .build();
+  }
+
+  /**
+   * Called when an element is evicted from the cache.
+   * @param path the path
+   * @param state the state
+   */
+  void onCacheEviction(String path, SyncState state) {
+    try {
+      // on eviction, we must mark our parent as needing a sync with our invalidation time
+      notifyInvalidationInternal(PathUtils.getParent(path), state.mInvalidationTime);
+    } catch (InvalidPathException e) {
+      throw new RuntimeException("Should not have an invalid path in the cache", e);
+    }
+  }
+
+  /**
+   * There is a most one writer at a time to a sync state.
+   * But multiple readers may be reading the values while they
+   * are being updated, thus the values are volatile.
+   */
+  static class SyncState {
+    volatile long mDirectChildInvalidation = 0;
+    volatile long mRecursiveChildInvalidation = 0;
+    volatile long mInvalidationTime = 0;
+    volatile long mValidationTime = 0;
+    volatile long mDirectValidationTime = 0;
+    volatile long mRecursiveValidationTime = 0;
```

Comment:
i might want to add a new one
mValidationFailedTime for the last timestamp for the last failed metadata sync. This will be used to prevent many metadata sync happening concurrently for the same path. I'll put up my PR later.

Reply:
I merged with your PR, but I changed the semantics slightly. Now a sync is skipped if the sync time has been updated to a larger value since the start of the sync.

I needed to do this because I changed how the success time of a sync is calculated. Before it used the time after the sync was completed, now it uses the time at the start of the sync. This is needed in case of concurrent invalidations coming in after the start of the sync and before the end of the sync. If the end of the sync time is used then this invalidations will be missed, so we need to use the start sync time.

Reply:
i think the ts won't be updated unless it succeeded

Reply:
yeah I think the approach in general works and looks good to me. Plus we've got some unit tests for it already. One of my concern is that we have this static variable

`  public static final SyncCheck SHOULD_SYNC = new SyncCheck(
      true, 0);
` 
and i'm wondering if this can be returned as a return value of the shouldSync() method or not? If yes, the condition check you added in InodeSyncStream might be broken. I don't see you are using this variable anywhere though... Just to double confirm.

Reply:
> If we use the start sync time, what happens if the sync fails (due to IOException or something else)? Do we roll the timestamp back?

Yes the timestamp is only updated when the sync is successful (there are actually some issues with this currently https://tachyonnexus.atlassian.net/browse/AE-2633, but this is being addressed in another PR).

Reply:
> yeah I think the approach in general works and looks good to me. Plus we've got some unit tests for it already. One of my concern is that we have this static variable
> 
> ` public static final SyncCheck SHOULD_SYNC = new SyncCheck( true, 0);` and i'm wondering if this can be returned as a return value of the shouldSync() method or not? If yes, the condition check you added in InodeSyncStream might be broken. I don't see you are using this variable anywhere though... Just to double confirm.

I changed it, to make sure the sync time isn't used in this case.